zkvm/src/constraints.rs

```diff
@@ -89,21 +92,64 @@ impl Constraint {
         .map_err(|e| VMError::R1CSError(e))
     }
 
-    fn flatten<CS: r1cs::RandomizedConstraintSystem>(self, cs: &mut CS) -> r1cs::LinearCombination {
+    fn flatten<CS: r1cs::RandomizedConstraintSystem>(
+        self,
+        cs: &mut CS,
+    ) -> Result<(r1cs::LinearCombination, Option<Scalar>), r1cs::R1CSError> {
         match self {
-            Constraint::Eq(expr1, expr2) => expr1.to_r1cs_lc() - expr2.to_r1cs_lc(),
+            Constraint::Eq(expr1, expr2) => {
+                let lc = expr1.to_r1cs_lc() - expr2.to_r1cs_lc();
+                match (expr1.eval(), expr2.eval()) {
+                    (Some(x), Some(y)) => Ok((lc, Some(x.to_scalar() - y.to_scalar()))),
+                    (_, _) => Ok((lc, None)),
+                }
+            }
             Constraint::And(c1, c2) => {
-                let a = c1.flatten(cs);
-                let b = c2.flatten(cs);
+                let (a, a_assg) = c1.flatten(cs)?;
+                let (b, b_assg) = c2.flatten(cs)?;
                 let z = cs.challenge_scalar(b"ZkVM.verify.and-challenge");
-                a + z * b
+                match (a_assg, b_assg) {
+                    (Some(a_assg), Some(b_assg)) => Ok((a + z * b, Some(a_assg + z * b_assg))),
```

Comment:
can we factor LC logic out of assignment logic? So assignment mapping is completely separate. 

e.g.

```rust
let assignment = match (a_assg, b_assg) {
    (Some(a), Some(b)) => Some(a + z * b),
    _ => None
};
(a + z*b, assignment)
```

Reply:
Alternatively:

```rust
let assignment = a_assg.and_then(|a| b_assg.map(|b| a + z * b ));
(a + z*b, assignment)
```

Reply:
til about `and_then`!